### PR TITLE
Ryhmanaytto

### PR DESCRIPTION
**Pull request by mkamka**

### CODE REVIEW

File: organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx [hunk 1]
```diff
@@ -52,51 +50,46 @@ const Ryhmat: React.FC = (props) => {
                     </a>
                 );
             },
-            collapse: true,
+            accessor: (values: any) => {
+                return values.nimi[language] || values.nimi.fi || values.nimi.sv || values.nimi.en || '';
```
Comment (jollem) [inline organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx:54]: This looks bit ugly. ok it might work now but what about if set of supported languages change for some reason? It _might_ make sense to extract this kind of logic to some common place

File: organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx [hunk 2]
```diff
@@ -52,51 +50,46 @@ const Ryhmat: React.FC = (props) => {
...
-                                    koodityyppi.nimi['fi'] ||
-                                    koodityyppi.nimi['sv'] ||
-                                    koodityyppi.nimi['en']
-                                );
-                            }
-                            return tyyppi;
-                        })
+                if (row.original.ryhmatyypit.length > 0) {
+                    return row.original.ryhmatyypit
+                        .map((tyyppi: string) => ryhmaTyypitKoodisto.uri2Nimi(tyyppi.slice(0, -2)))
```
Comment (jollem) [inline organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx:64]: slice? why? wtf? could be worked around by .... naming. for example dropSuffix(tyyppi). 
bonus: it could be tested easily

File: organisaatio-ui/src/components/Sivut/Ryhmat/__snapshots__/Ryhmat.test.tsx.snap [hunk 3]
```diff
@@ -0,0 +1,3 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`Ryhmat Rendering Renders without crashing 1`] = `<Component />`;
```
Comment (jollem) [inline organisaatio-ui/src/components/Sivut/Ryhmat/__snapshots__/Ryhmat.test.tsx.snap:3]: This would make more sense if rendered with some feasible properties

File: organisaatio-ui/src/components/Sivut/Tyypit/Tyypit.tsx [hunk 4]
```diff
@@ -29,7 +30,7 @@ const urlPrefix = process.env.NODE_ENV === 'development' ? '/api' : 'organisaati
 
 const Tyypit = (props: Props) => {
     const { i18n, language } = useContext(LanguageContext);
-    const [tyypit, setTyypit] = useState<string[] | YhteystietoTyyppi[] | undefined>(undefined);
+    const [tyypit, setTyypit] = useState<YhteystietoTyyppi[] | undefined>(undefined);
```
Comment (jollem) [inline organisaatio-ui/src/components/Sivut/Tyypit/Tyypit.tsx:33]: why accepting undefined? consider using empty array as default value

File: organisaatio-ui/src/components/Taulukot/NormaaliTaulukko.tsx [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+import { Ryhma, SelectOptionType, YhteystietoTyyppi } from '../../types/types';
+import { ValueType } from 'react-select';
+
+// TODO yhtenäistä!
+const mapPaginationSelectors = (index) => {
+    if (index < 3) return [0, 5];
+    return [index - 2, index + 3];
+};
+
+export type Props = {
```
Comment (jollem) [inline organisaatio-ui/src/components/Taulukot/NormaaliTaulukko.tsx:32]: rather lenient typing?
being more strict _will_ make life easier

File: organisaatio-ui/src/components/Taulukot/NormaaliTaulukko.tsx [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+    const { i18n } = useContext(LanguageContext);
+    const { ryhmaTyypitKoodisto, kayttoRyhmatKoodisto, ryhmanTilaKoodisto } = useContext(KoodistoContext);
+    const [tyyppiFiltteri, setTyyppiFiltteri] = useState<
+        ValueType<SelectOptionType>[] | ValueType<SelectOptionType> | undefined
+    >();
+    const [kayttoRyhmatFiltteri, setKayttoRyhmatFiltteri] = useState<ValueType<SelectOptionType>[] | undefined>();
+    const [tilaFiltteri, setTilaFiltteri] = useState<ValueType<SelectOptionType>[] | undefined>();
+
+    const ryhmatyypitOptions = ryhmaTyypitKoodisto.koodit().map((k) => {
+        return { value: k.arvo, label: k.nimi['fi'] || k.nimi['sv'] || k.nimi['en'] || '' };
```
Comment (jollem) [inline organisaatio-ui/src/components/Taulukot/NormaaliTaulukko.tsx:56]: ...and here we see that localization kludge again for label? seems to be repeating pattern. what if user prefers some other locale as default?

File: organisaatio-ui/src/components/Taulukot/NormaaliTaulukko.tsx [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+                            onChange={(
+                                values: ValueType<SelectOptionType>[] | ValueType<SelectOptionType> | undefined
+                            ) => {
+                                setKayttoRyhmatFiltteri(values as ValueType<SelectOptionType>[] | undefined);
+                                let kayttoryhmat: string | undefined = undefined;
+                                if (values && (values as ValueType<SelectOptionType>[]).length > 0) {
+                                    kayttoryhmat = (values as ValueType<SelectOptionType>[])
+                                        .map((val) => val && (val as SelectOptionType).label)
+                                        .join(', ');
+                                }
```
Comment (jollem) [inline organisaatio-ui/src/components/Taulukot/NormaaliTaulukko.tsx:113]: repetition from above?

File: organisaatio-ui/src/components/Taulukot/NormaaliTaulukko.tsx [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+    ryhmatData = [],
+    yhteystietoTyypitData = [],
+    ryhmatColumns = [],
+    yhteystietotyypitColumns = [],
+    useHakuFiltteri = false,
+}: Props) {
+    const { i18n } = useContext(LanguageContext);
+
+    let inputData = [] as any; // TODO check type casting bc i could not get | to work.
+    let tableColumns = [] as any;
```
Comment (jollem) [inline organisaatio-ui/src/components/Taulukot/NormaaliTaulukko.tsx:158]: hmmm.. any tells me nothing

File: organisaatio-ui/src/types/react-table-config.d.ts [hunk 9]
```diff
@@ -0,0 +1,62 @@
+import {
```
Comment (jollem) [inline organisaatio-ui/src/types/react-table-config.d.ts:1]: 😲
what is this file. why it is needed? where is it taken from? is it modified somehow? 
...dunno details BUT:
might make sense to add it as-is in separate commit with adequate comment. only after that make necessary modification in separate commit.
If there are no changes, do we need to have it in our own repository?

File: organisaatio-ui/src/types/types.ts [hunk 10]
```diff
@@ -78,8 +78,8 @@ export type Yhteystiedot = YhteystiedotEmail | YhteystiedotPhone | YhteystiedotW
 export interface Organisaatio {
     oid?: string;
     ytunnus: string;
-    nimi: any;
-    nimet: any;
+    nimi: any; // TODO?
```
Comment (jollem) [inline organisaatio-ui/src/types/types.ts:81]: TODO

File: organisaatio-ui/src/OrganisaatioApp.tsx [hunk 11]
```diff
@@ -60,35 +60,55 @@ const OrganisaatioApp: React.FC = () => {
...
+
     if (
         languageLoading ||
         lokalisointiLoading ||
         kunnatLoading ||
         ryhmaTyypitLoading ||
         kayttoRyhmatLoading ||
-        organisaatioTyypitLoading
+        organisaatioTyypitLoading ||
+        ryhmanTilatLoading
```
Comment (jollem) [inline organisaatio-ui/src/OrganisaatioApp.tsx:75]: not really mandatory but this could be written like:
```
if ( loading() ) {
    return spinner();
}
```

This would make it easier to:
- add new resources
- unit test spinner view

File: organisaatio-ui/src/OrganisaatioApp.tsx [hunk 12]
```diff
@@ -60,35 +60,55 @@ const OrganisaatioApp: React.FC = () => {
...
     ) {
         return (
             <ThemeProvider theme={theme}>
                 <Spin />
             </ThemeProvider>
         );
     }
-    if (lokalisointiError || kunnatError || ryhmaTyypitError || kayttoRyhmatError || organisaatioTyypitError) {
+    if (
```
Comment (jollem) [inline organisaatio-ui/src/OrganisaatioApp.tsx:83]:
```
if ( hasError) {
    return errorPage()
}
```

File: organisaatio-ui/src/components/Taulukot/Normaalitaulukko.test.tsx [hunk 13]
```diff
@@ -0,0 +1,50 @@
+import React from 'react';
```
Comment (jollem) [inline organisaatio-ui/src/components/Taulukot/Normaalitaulukko.test.tsx:1]: good starting point for adding more test... 
- feed in some props, see it they are transformed properly
- what happens on input events?
- ...so forth

**Review by jollem — APPROVED** [review]
approved BUT:
- please go through all comments. some things would be really easy to improve
- big bonus 👍 for tests (there are never enough)

File: organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx [hunk 14]
```diff
@@ -4,45 +4,43 @@ import TyypitJaRyhmatKehys from '../TyypitJaRyhmatKehys/TyypitJaRyhmatKehys';
...
 import Axios from 'axios';
 import { useState } from 'react';
 import { Ryhma } from '../../../types/types';
 import NormaaliTaulukko from '../../Taulukot/NormaaliTaulukko';
 import Spin from '@opetushallitus/virkailija-ui-components/Spin';
+import { Column } from 'react-table';
 
 const urlPrefix = process.env.NODE_ENV === 'development' ? '/api' : '/organisaatio';
 
 const Ryhmat: React.FC = (props) => {
```
Comment (jollem) [inline organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx:17]: what is "props" and what it is used for?

File: organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx [hunk 15]
```diff
@@ -4,45 +4,43 @@ import TyypitJaRyhmatKehys from '../TyypitJaRyhmatKehys/TyypitJaRyhmatKehys';
...
+    const { i18n, language } = useContext(LanguageContext);
     const { ryhmaTyypitKoodisto, kayttoRyhmatKoodisto } = useContext(KoodistoContext);
     const [ryhmat, setRyhmat] = useState<Ryhma[] | undefined>(undefined);
```
Comment (jollem) [inline organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx:20]: could be initialized to proper value instead of undefined which is rather dangerous

File: organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx [hunk 16]
```diff
@@ -4,45 +4,43 @@ import TyypitJaRyhmatKehys from '../TyypitJaRyhmatKehys/TyypitJaRyhmatKehys';
...
+
     useEffect(() => {
         async function fetch() {
```
Comment (jollem) [inline organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx:23]: stick either to fat arrow OR traditional function definition (i prefer fat arrow)

File: organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx [hunk 17]
```diff
@@ -4,45 +4,43 @@ import TyypitJaRyhmatKehys from '../TyypitJaRyhmatKehys/TyypitJaRyhmatKehys';
...
             try {
                 const response = await Axios.get(`${urlPrefix}/organisaatio/v3/ryhmat?aktiivinen=true`);
```
Comment (jollem) [inline organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx:25]: response type could / should be typed

Comment (jollem) [inline organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx:25]: there is mixed use of useAxios and Axios.get throughout the application

Comment (jollem) [inline organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx:25]: it would be very beneficial to abstract all api calls to separate module (and properly add type information for expected responses)

File: organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx [hunk 18]
```diff
@@ -4,45 +4,43 @@ import TyypitJaRyhmatKehys from '../TyypitJaRyhmatKehys/TyypitJaRyhmatKehys';
...
                 const data = response.data;
-                console.log('data', data);
                 setRyhmat([...data]);
             } catch (error) {
                 console.error('error fetching', error);
```
Comment (jollem) [inline organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx:29]: humm.. so perpetual spinner as error handling?

**Review by jollem — APPROVED** [review]
There are some details which might require more throughout renovation. Still accepting this feature as is because the renovation might require quite lot of effort.

File: organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx (hunk 2 above)
Comment (mkamka) [reply organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx:64]: This name versioning( when the version number actually is there and when it is not) is a bit of a mystery to me and hence the slicing for now. i will move it to a dropsuffix func as suggested. Thanks!

File: organisaatio-ui/src/components/Sivut/Ryhmat/__snapshots__/Ryhmat.test.tsx.snap (hunk 3 above)
Comment (mkamka) [reply organisaatio-ui/src/components/Sivut/Ryhmat/__snapshots__/Ryhmat.test.tsx.snap:3]: I think there is no props for now,if i will remove unused props from the component itself, will it be okay then?

File: organisaatio-ui/src/components/Sivut/Tyypit/Tyypit.tsx (hunk 4 above)
Comment (mkamka) [reply organisaatio-ui/src/components/Sivut/Tyypit/Tyypit.tsx:33]: i will try to change this accordingly

File: organisaatio-ui/src/components/Taulukot/NormaaliTaulukko.tsx (hunk 5 above)
Comment (mkamka) [reply organisaatio-ui/src/components/Taulukot/NormaaliTaulukko.tsx:32]: will change this to NormaaliTaulukkoProps or something similar

File: organisaatio-ui/src/types/types.ts (hunk 10 above)
Comment (mkamka) [reply organisaatio-ui/src/types/types.ts:81]: added todo because nimi handling is a bit of a mystery to me and will check it later.

File: organisaatio-ui/src/types/react-table-config.d.ts (hunk 9 above)
Comment (mkamka) [reply organisaatio-ui/src/types/react-table-config.d.ts:1]: I think is needed for React Table. Added as instucted here: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-table/Readme.md. I could not find another way to make this work with React Table.

File: organisaatio-ui/src/OrganisaatioApp.tsx (hunk 11 above)
Comment (mkamka) [reply organisaatio-ui/src/OrganisaatioApp.tsx:75]: i suggest making a separate ticket for this.

File: organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx (hunk 14 above)
Comment (mkamka) [reply organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx:17]: will remove props

File: organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx (hunk 15 above)
Comment (mkamka) [reply organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx:20]: will try to remove these undefined initialations

File: organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx (hunk 17 above)
Comment (mkamka) [reply organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx:25]: will try to type it

File: organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx (hunk 18 above)
Comment (mkamka) [reply organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx:29]: error handling is not properly implemented yet. Maybe this could be done at the same time?

Comment (jollem) [reply organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx:29]: yep, that the point. by isolating api access to one place, it's much more easier to implement something cross-cutting like error handling

File: organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx (hunk 1 above)
Comment (mkamka) [reply organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx:54]: Yes indeed. This nimikikkailu is a little bit of a mystery to me so there is plenty of things to be improved im sure.

File: organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx (hunk 16 above)
Comment (mkamka) [reply organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx:23]: wil change to arrow

File: organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx [hunk 19]
```diff
@@ -4,111 +4,118 @@ import TyypitJaRyhmatKehys from '../TyypitJaRyhmatKehys/TyypitJaRyhmatKehys';
...
+    return koodi;
+};
+
+const Ryhmat: React.FC = () => {
+    const { i18n, language } = useContext(LanguageContext);
     const { ryhmaTyypitKoodisto, kayttoRyhmatKoodisto } = useContext(KoodistoContext);
-    const [ryhmat, setRyhmat] = useState<Ryhma[] | undefined>(undefined);
+    const [ryhmat, setRyhmat] = useState<Ryhma[]>([]);
+    // TODO TS jutut ja haku mapperit tyyppien ja käyttötarkoituksen osalta
+    const RyhmatColumns: Column<Ryhma>[] = React.useMemo(
```
Comment (mkamka) [inline organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx:28]: Tämä muuttui

File: organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx (hunk 18 above)
Comment (mkamka) [reply organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx:109]: nämä muutettu

File: organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx [hunk 20]
```diff
@@ -4,111 +4,118 @@ import TyypitJaRyhmatKehys from '../TyypitJaRyhmatKehys/TyypitJaRyhmatKehys';
...
                 console.error('error fetching', error);
             }
-        }
+        };
         fetch();
     }, []);
-    const { i18n, language } = useContext(LanguageContext);
 
-    if (!ryhmat) {
+    if (ryhmat.length === 0) {
```
Comment (mkamka) [inline organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx:115]: tämä muuttui

File: organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx [hunk 21]
```diff
@@ -4,111 +4,118 @@ import TyypitJaRyhmatKehys from '../TyypitJaRyhmatKehys/TyypitJaRyhmatKehys';
...
-import Axios from 'axios';
+import Axios, { AxiosResponse } from 'axios';
 import { useState } from 'react';
 import { Ryhma } from '../../../types/types';
 import NormaaliTaulukko from '../../Taulukot/NormaaliTaulukko';
 import Spin from '@opetushallitus/virkailija-ui-components/Spin';
+import { Column } from 'react-table';
 
-const Ryhmat: React.FC = (props) => {
+export const dropKoodiVersionSuffix = (koodi: string) => {
```
Comment (mkamka) [inline organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx:15]: tämän lisäsin

File: organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx [hunk 22]
```diff
@@ -4,111 +4,118 @@ import TyypitJaRyhmatKehys from '../TyypitJaRyhmatKehys/TyypitJaRyhmatKehys';
...
+    const hasVersioningHashtag = koodi.search('#');
+    if (hasVersioningHashtag && hasVersioningHashtag) {
+        return koodi.slice(0, hasVersioningHashtag);
+    }
+    return koodi;
+};
+
+const Ryhmat: React.FC = () => {
```
Comment (mkamka) [inline organisaatio-ui/src/components/Sivut/Ryhmat/Ryhmat.tsx:23]: poistin propsin
